pkg/controller/jobframework/reconciler.go

```diff
@@ -537,6 +544,42 @@ func FindMatchingWorkloads(ctx context.Context, c client.Client, job GenericJob)
 	return match, toDelete, nil
 }
 
+func (r *JobReconciler) getPrebuiltWorkload(ctx context.Context, job GenericJob, object client.Object, name string) (*kueue.Workload, error) {
+	wl := &kueue.Workload{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: name, Namespace: object.GetNamespace()}, wl)
+	if err != nil {
+		return nil, client.IgnoreNotFound(err)
+	}
+
+	if !metav1.IsControlledBy(wl, object) {
+		if err := ctrl.SetControllerReference(object, wl, r.client.Scheme()); err != nil {
+			// don't return an error here, since a retry cannot give a different result
+			return nil, nil
+		}
+
+		if errs := validation.IsValidLabelValue(string(object.GetUID())); len(errs) == 0 {
+			wl.Labels = maps.MergeKeepFirst(map[string]string{controllerconsts.JobUIDLabel: string(object.GetUID())}, wl.Labels)
+		}
+
+		if err := r.client.Update(ctx, wl); err != nil {
+			return nil, err
+		}
+	}
+
+	if !equivalentToWorkload(job, wl) {
```

Comment:
Currently `getPrebuiltWorkload` may do mutating actions: "adopt" and "mark as finished". I would suggest splitting this function into two with the dedicated functions, proposing: `ensurePrebuildWorkloadIsAdopted` and `markPrebuildWorkloadFinishedIfOutOfSync` (or something along the lines).

Reply:
done